[FSDP] Generalize fsdp_modules()

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73716
* __->__ #73553

Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

fegin committed:
[FSDP] Generalize fsdp_modules()

Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
Lgtm but could we have some unit test with this change?

zhaojuanmao reviewed (commented):
codes look good to me, thanks for adding this!

fegin committed:
Update on "[FSDP] Generalize fsdp_modules()"


Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Generalize fsdp_modules()"


Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Generalize fsdp_modules()"


Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Generalize fsdp_modules()"


Generalize fsdp_modules():
   1. Make fsdp_modules() a static method so that it can be called even when users have no reference to the FSDP module -- it can be wrapped inside an non-FSDP module.
   2. Add root_only option to returns only the root FSDP module(s).

Differential Revision: [D34537408](https://our.internmc.facebook.com/intern/diff/D34537408/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks for adding the tests!
